File: Arrays/best_buy_sell_stock.py
from typing import List


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        maxProfit = 0
        minPrice = float('inf')
        for price in prices:
            if price < minPrice:
                minPrice = price
            elif price - minPrice > maxProfit:
                maxProfit = price - minPrice
        return maxProfit


sol = Solution()
res = sol.maxProfit([7, 1, 5, 3, 6, 4])
print(res)

File: Arrays/best_buy_sell_stock.cpp

#include <iostream>
#include <vector>
#include <climits>

class Solution {
public:
    int maxProfit(std::vector<int>& prices) {
        int mProfit = 0;
        int minPrice = INT_MAX;
        for (int i = 0; i < prices.size(); i++) {
            if (prices[i] < minPrice) {
                minPrice = prices[i];
            }
            else if (prices[i] - minPrice > mProfit) {
                mProfit = prices[i] - minPrice;
            }
        }
        return mProfit;
    }
};

int main() {
    Solution s;
    std::vector<int> prices = {7,1,5,3,6,4};
    std::cout << s.maxProfit(prices) << std::endl;
    return 0;
}
